### PR TITLE
handle both engine- and factory-based tensors

### PR DESCRIPTION
**Pull request by arnej27959**
* use EngineOrFactory::get() instead of DefaultTensorEngine::ref()
* avoid direct use of DenseTensorView etc where possible
* use eval::Value instead of tensor::Tensor where possible

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review
@geirst please review
@toregge FYI
@baldersheim FYI
@vekterli FYI

### CODE REVIEW

File: searchcore/src/tests/proton/common/attribute_updater/attribute_updater_test.cpp [hunk 1]
```diff
@@ -442,7 +442,9 @@ struct TensorFixture : public Fixture {
     }
 
     void assertTensor(const TensorSpec &expSpec) {
-        EXPECT_EQUAL(expSpec, attribute->getTensor(1)->toSpec());
+        auto engine = EngineOrFactory::get();
+        auto actual = engine.to_spec(*attribute->getTensor(1));
```
Comment (havardpe) [inline searchcore/src/tests/proton/common/attribute_updater/attribute_updater_test.cpp:446]: combine lines?

File: searchcore/src/tests/proton/docsummary/docsummary.cpp [hunk 2]
```diff
@@ -139,9 +138,9 @@ getDocTypeName()
     return "searchdocument";
 }
 
-Tensor::UP make_tensor(const TensorSpec &spec) {
-    auto tensor = DefaultTensorEngine::ref().from_spec(spec);
-    return Tensor::UP(dynamic_cast<Tensor*>(tensor.release()));
+vespalib::eval::Value::UP make_tensor(const TensorSpec &spec) {
+    auto tensor = EngineOrFactory::get().from_spec(spec);
```
Comment (havardpe) [inline searchcore/src/tests/proton/docsummary/docsummary.cpp:142]: just return here?

File: searchcore/src/vespa/searchcore/proton/matching/docsum_matcher.cpp [hunk 3]
```diff
@@ -71,9 +72,9 @@ get_feature_set(const MatchToolsFactory &mtf,
             for (uint32_t j = 0; j < featureNames.size(); ++j) {
                 if (resolver.is_object(j)) {
                     auto obj = resolver.resolve(j).as_object(docId);
-                    if (const auto *tensor = obj.get().as_tensor()) {
+                    if (! obj.get().is_double()) {
```
Comment (havardpe) [inline searchcore/src/vespa/searchcore/proton/matching/docsum_matcher.cpp:75]: prefer checking type instead `obj.get().type().is_double()`

File: searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp [hunk 4]
```diff
@@ -50,36 +50,19 @@ using search::tensor::PrepareResult;
...
-    auto value = DefaultTensorEngine::ref().from_spec(spec);
-    if (value->is_double()) {
-        return Tensor::UP(new DenseTensor<double>(ValueType::double_type(), {value->as_double()}));
-    }
-    Tensor *tensor = dynamic_cast<Tensor*>(value.get());
-    ASSERT_TRUE(tensor != nullptr);
-    value.release();
-    return Tensor::UP(tensor);
+Value::UP createTensor(const TensorSpec &spec) {
+    auto value = EngineOrFactory::get().from_spec(spec);
```
Comment (havardpe) [inline searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp:64]: return here?

File: searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp [hunk 5]
```diff
@@ -922,12 +903,9 @@ class NearestNeighborBlueprintFixture : public DenseTensorAttributeMockIndex {
...
     }
 
-    std::unique_ptr<QueryTensor> createDenseTensor(const TensorSpec &spec) {
-        auto value = DefaultTensorEngine::ref().from_spec(spec);
-        QueryTensor *tensor = dynamic_cast<QueryTensor *>(value.get());
-        ASSERT_TRUE(tensor != nullptr);
-        value.release();
-        return std::unique_ptr<QueryTensor>(tensor);
+    std::unique_ptr<Value> createDenseTensor(const TensorSpec &spec) {
+        auto value = EngineOrFactory::get().from_spec(spec);
```
Comment (havardpe) [inline searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp:907]: return here?

File: searchlib/src/tests/features/tensor_from_weighted_set/tensor_from_weighted_set_test.cpp [hunk 6]
```diff
@@ -27,21 +28,20 @@ using search::StringAttribute;
...
 typedef search::attribute::BasicType AVBT;
 typedef search::attribute::CollectionType AVCT;
 typedef search::AttributeVector::SP AttributePtr;
 typedef FtTestApp FTA;
 
-Tensor::UP make_tensor(const TensorSpec &spec) {
-    auto tensor = DefaultTensorEngine::ref().from_spec(spec);
-    return Tensor::UP(dynamic_cast<Tensor*>(tensor.release()));
+Value::UP make_tensor(const TensorSpec &spec) {
+    auto tensor = EngineOrFactory::get().from_spec(spec);
```
Comment (havardpe) [inline searchlib/src/tests/features/tensor_from_weighted_set/tensor_from_weighted_set_test.cpp:40]: return here?

File: searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp [hunk 7]
```diff
@@ -3,26 +3,28 @@
...
 
 using EntryRef = DenseTensorStore::EntryRef;
 
-Tensor::UP
+Value::UP
 makeTensor(const TensorSpec &spec)
 {
-    auto tensor = DefaultTensorEngine::ref().from_spec(spec);
-    return Tensor::UP(dynamic_cast<Tensor *>(tensor.release()));
+    auto tensor = EngineOrFactory::get().from_spec(spec);
```
Comment (havardpe) [inline searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp:26]: return here?

File: searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp [hunk 8]
```diff
@@ -724,19 +723,17 @@ class CreateBlueprintVisitor : public CreateBlueprintVisitorHelper
...
+        const auto & qt_type = query_tensor->type();
+        if (! qt_type.is_dense()) {
             return fail_nearest_neighbor_term(n, make_string("Query tensor is not a dense tensor (type=%s)",
                                                              query_tensor->type().to_spec().c_str()));
         }
-        if (!is_compatible_for_nearest_neighbor(dense_attr_tensor->getTensorType(), dense_query_tensor->type())) {
+        if (!is_compatible_for_nearest_neighbor(dense_attr_tensor->getTensorType(), qt_type)) {
             return fail_nearest_neighbor_term(n, make_string("Attribute tensor type (%s) and query tensor type (%s) are not compatible",
-                                                             dense_attr_tensor->getTensorType().to_spec().c_str(), dense_query_tensor->type().to_spec().c_str()));
+                                                             dense_attr_tensor->getTensorType().to_spec().c_str(), query_tensor->type().to_spec().c_str()));
```
Comment (havardpe) [inline searchlib/src/vespa/searchlib/attribute/attribute_blueprint_factory.cpp:733]: query_tensor->type() -> qt_type

**Review by havardpe — COMMENTED** [review]
This looks very good

I suggest we try not to use the functions Value::is_double or Value::is_tensor, but rather check the type instead. It would also be a good smoke test to run all unit tests with FastValueBuilderFactory as default implementation (temporary one-line change that is not to be commited)

File: document/src/vespa/document/update/tensor_add_update.cpp [hunk 9]
```diff
@@ -78,14 +82,34 @@ TensorAddUpdate::checkCompatibility(const Field& field) const
...
-        return tensor.add(*addTensor);
+        auto engine = EngineOrFactory::get();
+        if (engine.is_factory()) {
+            return TensorPartialUpdate::add(tensor, *addTensor, engine.factory());
+        } else {
+            return old_add(&tensor, addTensor);
+        }
     }
-    return std::unique_ptr<Tensor>();
+    return std::unique_ptr<vespalib::eval::Value>();
```
Comment (geirst) [inline document/src/vespa/document/update/tensor_add_update.cpp:112]: Consider `return {};` instead.

File: document/src/vespa/document/update/tensor_modify_update.cpp [hunk 10]
```diff
@@ -156,16 +160,32 @@ TensorModifyUpdate::checkCompatibility(const Field& field) const
...
-        return tensor.modify(getJoinFunction(_operation), cellValues);
+        auto engine = EngineOrFactory::get();
+        if (engine.is_factory()) {
+            return TensorPartialUpdate::modify(tensor, getJoinFunction(_operation), *cellsTensor, engine.factory());
+        } else {
+            return old_modify(&tensor, cellsTensor, getJoinFunction(_operation));
+        }
     }
-    return std::unique_ptr<Tensor>();
+    return std::unique_ptr<vespalib::eval::Value>();
```
Comment (geirst) [inline document/src/vespa/document/update/tensor_modify_update.cpp:188]: Consider `return {};` instead.

File: document/src/vespa/document/update/tensor_remove_update.cpp [hunk 11]
```diff
@@ -102,16 +117,19 @@ TensorRemoveUpdate::checkCompatibility(const Field &field) const
...
-        return tensor.remove(cellAddresses);
+        auto engine = EngineOrFactory::get();
+        if (engine.is_factory()) {
+            return TensorPartialUpdate::remove(tensor, *addressTensor, engine.factory());
+        } else {
+            return old_remove(&tensor, addressTensor);
+        }
     }
-    return std::unique_ptr<Tensor>();
+    return std::unique_ptr<vespalib::eval::Value>();
```
Comment (geirst) [inline document/src/vespa/document/update/tensor_remove_update.cpp:132]: Consider `return {};` instead.

File: searchcore/src/tests/proton/docsummary/docsummary.cpp [hunk 12]
```diff
@@ -335,7 +334,7 @@ assertTensor(const Tensor::UP & exp, const std::string & fieldName,
     EXPECT_EQUAL(exp.get() == nullptr, data.size == 0u);
     if (exp) {
         vespalib::nbostream x(data.data, data.size);
-        Tensor::UP tensor = vespalib::tensor::TypedBinaryFormat::deserialize(x);
+        vespalib::eval::Value::UP tensor = EngineOrFactory::get().decode(x);
```
Comment (geirst) [inline searchcore/src/tests/proton/docsummary/docsummary.cpp:337]: Consider `auto tensor` instead.

File: searchcore/src/tests/proton/docsummary/summaryfieldconverter_test.cpp [hunk 13]
```diff
@@ -675,9 +676,9 @@ Test::requireThatPredicateIsPrinted()
                 SFC::convertSummaryField(false, *doc.getValue("predicate")).get());
 }
 
-Tensor::UP make_tensor(const TensorSpec &spec) {
-    auto tensor = DefaultTensorEngine::ref().from_spec(spec);
-    return Tensor::UP(dynamic_cast<Tensor*>(tensor.release()));
+Value::UP make_tensor(const TensorSpec &spec) {
+    auto tensor = EngineOrFactory::get().from_spec(spec);
```
Comment (geirst) [inline searchcore/src/tests/proton/docsummary/summaryfieldconverter_test.cpp:680]: Consider returning here.

File: searchcore/src/tests/proton/common/attribute_updater/attribute_updater_test.cpp [hunk 14]
```diff
@@ -408,11 +408,11 @@ getTensorDataType(const vespalib::string &spec)
     return *insres.first->second;
 }
 
-std::unique_ptr<Tensor>
+std::unique_ptr<Value>
 makeTensor(const TensorSpec &spec)
 {
-    auto result = DefaultTensorEngine::ref().from_spec(spec);
-    return std::unique_ptr<Tensor>(dynamic_cast<Tensor*>(result.release()));
+    auto result = EngineOrFactory::get().from_spec(spec);
```
Comment (geirst) [inline searchcore/src/tests/proton/common/attribute_updater/attribute_updater_test.cpp:414]: Consider returning here.

File: searchcore/src/tests/proton/server/documentretriever_test.cpp [hunk 15]
```diff
@@ -110,8 +113,8 @@ const char dyn_field_nas[] = "dynamic null attr string field"; // in document, n
 const char position_field[] = "position_field";
 vespalib::string dyn_field_tensor("dynamic_tensor_field");
 vespalib::string tensor_spec("tensor(x{})");
-std::unique_ptr<Tensor> static_tensor = makeTensor<Tensor>(TensorSpec(tensor_spec).add({{"x", "1"}}, 1.5));
-std::unique_ptr<Tensor> dynamic_tensor = makeTensor<Tensor>(TensorSpec(tensor_spec).add({{"x", "2"}}, 3.5));
+std::unique_ptr<Value> static_tensor = makeTensor<Value>(TensorSpec(tensor_spec).add({{"x", "1"}}, 1.5));
```
Comment (geirst) [inline searchcore/src/tests/proton/server/documentretriever_test.cpp:116]: Consider using auto.

File: searchcore/src/tests/proton/server/documentretriever_test.cpp [hunk 16]
```diff
@@ -110,8 +113,8 @@ const char dyn_field_nas[] = "dynamic null attr string field"; // in document, n
...
+std::unique_ptr<Value> dynamic_tensor = makeTensor<Value>(TensorSpec(tensor_spec).add({{"x", "2"}}, 3.5));
```
Comment (geirst) [inline searchcore/src/tests/proton/server/documentretriever_test.cpp:117]: Consider using auto.

File: searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp [hunk 17]
```diff
@@ -435,14 +418,14 @@ struct Fixture {
 
     void assertGetNoTensor(uint32_t docId) {
         AttributeGuard guard(_attr);
-        Tensor::UP actTensor = _tensorAttr->getTensor(docId);
+        Value::UP actTensor = _tensorAttr->getTensor(docId);
```
Comment (geirst) [inline searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp:421]: Consider using auto.

File: searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp [hunk 18]
```diff
@@ -435,14 +418,14 @@ struct Fixture {
...
         EXPECT_FALSE(actTensor);
     }
 
     void assertGetTensor(const TensorSpec &expSpec, uint32_t docId) {
-        Tensor::UP expTensor = createTensor(expSpec);
+        Value::UP expTensor = createTensor(expSpec);
```
Comment (geirst) [inline searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp:426]: Consider using auto.

File: searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp [hunk 19]
```diff
@@ -435,14 +418,14 @@ struct Fixture {
...
         AttributeGuard guard(_attr);
-        Tensor::UP actTensor = _tensorAttr->getTensor(docId);
+        Value::UP actTensor = _tensorAttr->getTensor(docId);
```
Comment (geirst) [inline searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp:428]: Consider using auto.

File: searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp [hunk 20]
```diff
@@ -655,7 +638,7 @@ void
 Fixture::testEmptyTensor()
 {
     const TensorAttribute &tensorAttr = *_tensorAttr;
-    Tensor::UP emptyTensor = tensorAttr.getEmptyTensor();
+    Value::UP emptyTensor = tensorAttr.getEmptyTensor();
```
Comment (geirst) [inline searchlib/src/tests/attribute/tensorattribute/tensorattribute_test.cpp:641]: Consider using auto.

File: searchlib/src/tests/features/constant/constant_test.cpp [hunk 21]
```diff
@@ -63,7 +62,7 @@ struct ExecFixture
     void addTensor(const vespalib::string &name,
                    const TensorSpec &spec)
     {
-        Tensor::UP tensor = make_tensor(spec);
+        Value::UP tensor = make_tensor(spec);
```
Comment (geirst) [inline searchlib/src/tests/features/constant/constant_test.cpp:65]: Consider using auto.

File: searchlib/src/tests/features/tensor_from_labels/tensor_from_labels_test.cpp [hunk 22]
```diff
@@ -26,21 +27,20 @@ using search::StringAttribute;
...
 typedef search::attribute::BasicType AVBT;
 typedef search::attribute::CollectionType AVCT;
 typedef search::AttributeVector::SP AttributePtr;
 typedef FtTestApp FTA;
 
-Tensor::UP make_tensor(const TensorSpec &spec) {
-    auto tensor = DefaultTensorEngine::ref().from_spec(spec);
-    return Tensor::UP(dynamic_cast<Tensor*>(tensor.release()));
+Value::UP make_tensor(const TensorSpec &spec) {
+    auto tensor = EngineOrFactory::get().from_spec(spec);
```
Comment (geirst) [inline searchlib/src/tests/features/tensor_from_labels/tensor_from_labels_test.cpp:39]: Consider returning here.

File: searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp [hunk 23]
```diff
@@ -32,23 +34,23 @@ struct Fixture
         : store(ValueType::from_spec(tensorType))
     {}
     void assertSetAndGetTensor(const TensorSpec &tensorSpec) {
-        Tensor::UP expTensor = makeTensor(tensorSpec);
+        Value::UP expTensor = makeTensor(tensorSpec);
```
Comment (geirst) [inline searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp:37]: Consider using auto.

File: searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp [hunk 24]
```diff
@@ -32,23 +34,23 @@ struct Fixture
...
         EntryRef ref = store.setTensor(*expTensor);
-        Tensor::UP actTensor = store.getTensor(ref);
-        EXPECT_EQUAL(expTensor->toSpec(), actTensor->toSpec());
+        Value::UP actTensor = store.getTensor(ref);
```
Comment (geirst) [inline searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp:39]: Consider using auto.

File: searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp [hunk 25]
```diff
@@ -32,23 +34,23 @@ struct Fixture
...
+        EXPECT_EQUAL(*expTensor, *actTensor);
         assertTensorView(ref, *expTensor);
     }
     void assertEmptyTensor(const TensorSpec &tensorSpec) {
-        Tensor::UP expTensor = makeTensor(tensorSpec);
+        Value::UP expTensor = makeTensor(tensorSpec);
```
Comment (geirst) [inline searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp:44]: Consider using auto.

File: searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp [hunk 26]
```diff
@@ -32,23 +34,23 @@ struct Fixture
...
         EntryRef ref;
-        Tensor::UP actTensor = store.getTensor(ref);
+        Value::UP actTensor = store.getTensor(ref);
```
Comment (geirst) [inline searchlib/src/tests/tensor/dense_tensor_store/dense_tensor_store_test.cpp:46]: Consider using auto.

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp [hunk 27]
```diff
@@ -164,15 +160,15 @@ DenseTensorAttribute::complete_set_tensor(DocId docid, const Tensor& tensor,
...
+std::unique_ptr<vespalib::eval::Value>
 DenseTensorAttribute::getTensor(DocId docId) const
 {
     EntryRef ref;
     if (docId < getCommittedDocIdLimit()) {
         ref = _refVector[docId];
     }
     if (!ref.valid()) {
-        return std::unique_ptr<Tensor>();
+        return std::unique_ptr<vespalib::eval::Value>();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/dense_tensor_attribute.cpp:171]: Consider `return {};` instead.

File: searchlib/src/vespa/searchlib/tensor/dense_tensor_store.cpp [hunk 28]
```diff
@@ -131,14 +127,14 @@ DenseTensorStore::move(EntryRef ref)
     return newraw.ref;
 }
 
-std::unique_ptr<Tensor>
+std::unique_ptr<Value>
 DenseTensorStore::getTensor(EntryRef ref) const
 {
     if (!ref.valid()) {
-        return std::unique_ptr<Tensor>();
+        return std::unique_ptr<Value>();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/dense_tensor_store.cpp:134]: Consider `return {};` instead.

File: searchlib/src/vespa/searchlib/tensor/serialized_tensor_attribute.cpp [hunk 29]
```diff
@@ -34,23 +34,23 @@ SerializedTensorAttribute::~SerializedTensorAttribute()
...
+std::unique_ptr<Value>
 SerializedTensorAttribute::getTensor(DocId docId) const
 {
     EntryRef ref;
     if (docId < getCommittedDocIdLimit()) {
         ref = _refVector[docId];
     }
     if (!ref.valid()) {
-        return std::unique_ptr<Tensor>();
+        return std::unique_ptr<Value>();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/serialized_tensor_attribute.cpp:53]: Consider `return {}; ` instead.

File: searchlib/src/vespa/searchlib/tensor/serialized_tensor_store.cpp [hunk 30]
```diff
@@ -87,21 +86,21 @@ SerializedTensorStore::move(EntryRef ref)
...
 }
 
-std::unique_ptr<Tensor>
+std::unique_ptr<Value>
 SerializedTensorStore::getTensor(EntryRef ref) const
 {
     auto raw = getRawBuffer(ref);
     if (raw.second == 0u) {
-        return std::unique_ptr<Tensor>();
+        return std::unique_ptr<Value>();
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/serialized_tensor_store.cpp:94]: Consider `return {};` instead.

**Review by geirst — APPROVED** [review]
Very nice 👍

File: searchlib/src/tests/queryeval/nearest_neighbor/nearest_neighbor_test.cpp [hunk 31]
```diff
@@ -42,7 +42,6 @@ DistanceFunction::UP euclid_f = search::tensor::make_distance_function(DistanceM
 
 std::unique_ptr<Value> createTensor(const TensorSpec &spec) {
     auto value = EngineOrFactory::get().from_spec(spec);
```
Comment (havardpe) [inline searchlib/src/tests/queryeval/nearest_neighbor/nearest_neighbor_test.cpp:44]: can probably just return here

**Comment by geirst** [comment]
👍

File: searchlib/src/tests/features/tensor_from_weighted_set/tensor_from_weighted_set_test.cpp [hunk 32]
```diff
@@ -122,7 +121,6 @@ struct ExecFixture
     }
     const Value &extractTensor(uint32_t docid) {
         Value::CREF value = test.resolveObjectFeature(docid);
```
Comment (havardpe) [inline searchlib/src/tests/features/tensor_from_weighted_set/tensor_from_weighted_set_test.cpp:123]: can probably just return here using implicit cast operator from std::reference_wrapper<>

File: searchlib/src/tests/features/tensor_from_labels/tensor_from_labels_test.cpp [hunk 33]
```diff
@@ -123,7 +122,6 @@ struct ExecFixture
     }
     const Value &extractTensor(uint32_t docid) {
         Value::CREF value = test.resolveObjectFeature(docid);
```
Comment (havardpe) [inline searchlib/src/tests/features/tensor_from_labels/tensor_from_labels_test.cpp:124]: can probably just return here using implicit cast operator from std::reference_wrapper<>

**Review by havardpe — APPROVED** [review]

**Pull request merged**